client/src/partials/vouchers/complex.js

```diff
@@ -0,0 +1,261 @@
+  'CurrencyService', 'SessionService', 'FindEntityService',
+  'FindReferenceService', 'NotifyService'
+];
+
+/**
+ * Complex Journal Vouchers
+ *
+ * This module implements complex journal vouchers. It allows users to quickly create transactions by
+ * specifying two or more lines of transactions and all relative document references
+ *
+ * @constructor
+ *
+ * @todo - Implement caching mechanism for incomplete forms (via AppCache)
+ * @todo - Implement Patient Invoices data and Cash Payment data for modal
+ * @todo - Implement a mean to categorise transactions for cashflow reports 
+ */
+function ComplexJournalVoucherController(Vouchers, $translate, Accounts, Currencies, Session, FindEntity, FindReference, Notify) {
+  var vm = this;
+
+  // bread crumb paths
+  vm.paths = [{
+    label : $translate.instant('VOUCHERS.COMPLEX.TITLE'),
+    current: true
+  }];
+
+  // bind the startup method as a reset method
+  vm.submit = submit;
+  vm.currencySymbol     = currencySymbol;
+  vm.addVoucherItem     = addVoucherItem;
+  vm.removeVoucherItem  = removeVoucherItem;
+  vm.checkRowValidity   = checkRowValidity;
+  vm.selectAccount      = selectAccount;
+  vm.openEntityModal    = FindEntity.openModal;
+  vm.openReferenceModal = FindReference.openModal;
+
+  // load the list of accounts
+  Accounts.read()
+  .then(function (accounts) {
+    vm.accounts = accounts.map(function (item) {
+      // remove children accounts
+      delete item.children;
+      return item;
+    });
+  });
+
+  // load the available currencies
+  Currencies.read()
+  .then(function (currencies) {
+    currencies.forEach(function (currency) {
+      currency.label = Currencies.format(currency.id);
+    });
+    vm.currencies = currencies;
+  });
+
+  /** Get the selected currency symbol */
+  function currencySymbol(currency_id) {
+    if (!currency_id) { return ; }
+    return Currencies.symbol(currency_id);
+  }
+
+  /** Add transaction row */
+  function addVoucherItem() {
+    vm.rows.push(generateRow());
+    checkRowValidity();
+  }
+
+  /** remove transaction row */
+  function removeVoucherItem(index) {
+    vm.rows.splice(index, 1);
+    checkRowValidity();
+  }
+
+  /** generate row element */
+  function generateRow() {
+    return {
+      account_id    : undefined,
+      debit         : 0,
+      credit        : 0,
+      document_uuid : undefined,
+      entity_uuid   : undefined
+    };
+  }
+
+  /** clean and generate voucher correct data */
+  function handleVoucher() {
+    var voucher;
+    if (vm.sumCredit === vm.sumDebit) {
+      voucher = {
+        project_id : Session.project.id,
+        date : vm.voucher.date,
+        description : vm.voucher.description,
+        currency_id : vm.voucher.currency_id,
+        amount : vm.sumDebit,
+        user_id : Session.user.id,
+      };
+    }
+    return voucher;
+  }
+
+  /** clean and generate voucher items data */
+  function handleVoucherItems() {
+    var voucherItems = [];
+    if (vm.validInput) {
```

Comment:
Why do you need to check `vm.validInput` here?  It would be better to move this check out of the function so that the function has one concern: formatting the voucher_items to submit to the server.